libs/test-common/src/main/java/org/codice/ddf/test/common/mockito/PrivilegedVerificationMode.java

```diff
@@ -0,0 +1,142 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.test.common.mockito;
+
+import static org.mockito.Mockito.times;
+
+import com.google.common.annotations.Beta;
+import java.lang.reflect.Field;
+import java.security.AccessController;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Stream;
+import org.mockito.exceptions.base.MockitoAssertionError;
+import org.mockito.internal.verification.VerificationModeFactory;
+import org.mockito.internal.verification.api.VerificationData;
+import org.mockito.internal.verification.api.VerificationDataInOrder;
+import org.mockito.internal.verification.api.VerificationInOrderMode;
+import org.mockito.invocation.Invocation;
+import org.mockito.invocation.Location;
+import org.mockito.invocation.MatchableInvocation;
+import org.mockito.verification.VerificationMode;
+
+/**
+ * Class that extends Mockito's {@link VerificationMode} to verify that a mocked method has been
+ * called within a {@code AccessController#doPrivileged()} block.
+ *
+ * <p><b> This code is experimental. While this class is functional and tested, it may change or be
+ * removed in a future version of the library. </b>
+ */
+@Beta
+public class PrivilegedVerificationMode implements VerificationMode, VerificationInOrderMode {
+
+  private final VerificationMode mode;
+
+  private PrivilegedVerificationMode(VerificationMode mode) {
+    this.mode = mode;
+  }
+
+  /**
+   * Wraps another {@link VerificationMode} instance and ensures all the expected calls were made
+   * inside a {@code AccessController.doPrivileged()} block.
+   *
+   * @param mode {@link VerificationMode} instance to wrap with a privileged check
+   * @return wrapped {@link VerificationMode} instance
+   */
+  public static VerificationMode privileged(VerificationMode mode) {
+    return new PrivilegedVerificationMode(mode);
+  }
+
+  /**
+   * Returns a {@link VerificationMode} object that ensures a single expected calls was made inside
+   * a {@code AccessController.doPrivileged()} block.
+   *
+   * @return wrapped {@link VerificationMode} instance
+   */
+  public static VerificationMode privileged() {
+    return new PrivilegedVerificationMode(times(1));
+  }
+
+  @Override
+  public void verify(VerificationData verificationData) {
+    String target = verificationData.getTarget().toString();
+
+    mode.verify(verificationData);
+
+    verify(verificationData.getTarget(), verificationData.getAllInvocations());
+  }
+
+  @Override
+  public void verifyInOrder(VerificationDataInOrder verificationDataInOrder) {
+    String target = verificationDataInOrder.getWanted().toString();
+
+    if (!(mode instanceof VerificationInOrderMode)) {
```

Comment:
❓ Should we split this class in 2 such that we don't mix the `in order` mode with the normal one and catch this type of issue at compile time instead?

Reply:
I tried doing that initially but because of the way they have the interfaces and classes structured, having two separate classes didn't work very well. It also forced me to create a third class to share the common code which felt overkill.

But what convinced me to stop trying is that this is how they implement their own classes (e.g., https://github.com/mockito/mockito/blob/release/2.x/src/main/java/org/mockito/internal/verification/Times.java#L20).